tests/federation/test_federation_sender.py

```diff
@@ -370,6 +370,43 @@ def test_unreachable_server(self):
         devices = {edu["content"]["device_id"] for edu in self.edus}
         self.assertEqual({"D1", "D2", "D3"}, devices)
 
+    def test_prune_outbound_device_pokes(self):
+        """If a destination is unreachable, and the updates are pruned, we should get
+        a single update"""
+        mock_send_txn = self.hs.get_federation_transport_client().send_transaction
+        mock_send_txn.side_effect = lambda t, cb: defer.fail("fail")
+
+        # create devices
+        u1 = self.register_user("user", "pass")
+        self.login("user", "pass", device_id="D1")
+        self.login("user", "pass", device_id="D2")
+        self.login("user", "pass", device_id="D3")
+
+        # delete them again
+        self.get_success(
+            self.hs.get_device_handler().delete_devices(u1, ["D1", "D2", "D3"])
+        )
+
+        self.assertGreaterEqual(mock_send_txn.call_count, 4)
+
+        # run the prune job
+        self.reactor.advance(10)
+        self.get_success(
+            self.hs.get_datastore()._prune_old_outbound_device_pokes(prune_age=1)
+        )
+
+        # recover the server
+        mock_send_txn.side_effect = self.record_transaction
+        self.hs.get_federation_sender().send_device_messages("host2")
+        self.pump()
+
+        # there should be a single update for this user.
+        self.assertEqual(len(self.edus), 1)
+        edu = self.edus.pop(0)
+        self.assertEqual(edu["edu_type"], "m.device_list_update")
+        c = edu["content"]
+        self.assertEqual(c["prev_id"], [])
```

Comment:
Can we get a test where we successfully send an update before we prune please? Now that we're not deleting from `device_lists_outbound_last_success` its not clear what the `prev_id` will be in that case.

Separately, the spec doesn't actually say that the remote should resync when `prev_id` is empty, though Synapse does. We should either a) fix Synapse to point to a dropped update or b) update the spec. I guess I'm happy with not trying to fix it in this PR, but if so we should at least file an issue or something

Reply:
Yeah, thanks for making me think about this properly. Removing the deletion from `device_lists_outbound_last_success` was the wrong thing.

Fixing it isn't entirely trivial: have raised #7173 to track